Add TableProvider implementation with predicate pushdown and column p…

ShimonSte opened the pull request:
## Add TableProvider for Format-Based API Access #470 

Implements `ClickHouseTableProvider` to enable `.format("clickhouse")` syntax, essential for **Databricks Unity Catalog** environments where custom catalogs cannot be registered.

### Features

#### Core Functionality
- **Format name**: `"clickhouse"` - standard Spark DataSource V2 format registration
- **Predicate pushdown** & **column pruning** - verified via `system.query_log` inspection
- **Auto schema inference** - reads table schema from ClickHouse with ORDER BY generation
- **Nullable key support** - automatic `allow_nullable_key` setting for partition/sorting keys

#### Write Modes
- `SaveMode.Append` - append data to existing table
- `SaveMode.Overwrite` - truncate table before writing (full table overwrite) 

#### Overwrite Implementation
- Implements `SupportsOverwrite` trait in `ClickHouseWriteBuilder`  #33 
- Executes `TRUNCATE TABLE` before writing in overwrite mode
- Optimized distributed table handling with single `ON CLUSTER` command
- **Note**: Dynamic partition overwrite not yet implemented - currently only full table truncate

#### ClickHouse Cloud Compatibility
- **SharedMergeTree support** - added `SharedMergeTree` engine parsing in `AstVisitor`
- **Partition expression filtering** - filters unsupported Spark transformations (e.g., `xxHash64` in Spark 3.3) for SharedMergeTree tables
- **Graceful degradation** - keeps supported partition expressions while filtering incompatible ones
- Maintains Spark-side partitioning benefits while avoiding `AnalysisException` errors
- Applied across all Spark versions (3.3, 3.4, 3.5, 4.0)

### Usage

```scala
// Read with format API
val df = spark.read.format("clickhouse")
  .option("host", "localhost")
  .option("database", "default")
  .option("table", "my_table")
  .load()

// Write with overwrite mode
df.write.format("clickhouse")
  .mode(SaveMode.Overwrite)
  .option("host", "localhost")
  .option("database", "default")
  .option("table", "my_table")
  .save()

// Catalog API still available
spark.table("clickhouse.default.my_table")

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables format-based access via DataSource V2 for ClickHouse with improved write semantics and engine compatibility.
> 
> - Adds `ClickHouseTableProvider` (format name: `"clickhouse"`) with schema inference, predicate pushdown, and column pruning; registers via `META-INF/services/...DataSourceRegister`
> - Implements overwrite by adding `SupportsOverwrite` to `ClickHouseWrite` and truncating target tables (incl. `ON CLUSTER` for distributed) before write
> - Extends table capabilities with `OVERWRITE_BY_FILTER` and adds SharedMergeTree parsing in `AstVisitor`
> - Filters unsupported partition expressions for `SharedMergeTree` in `WriteJobDescription` to avoid analysis errors
> - Updates/extends IT tests: new `ClickHouseTableProviderSuite`; reader tests compare Catalog vs TableProvider paths; helper API updated to pass `(db, tbl)`
> - Build/doc updates: new `SPARK_VERSION_DIFFERENCES.md`; service files added; Gradle tweaks (mirror repo, ASM 9.6)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8b162bf21e1f79ea35d3da54ae165e1f705f328f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ShimonSte committed:
Add TableProvider implementation with predicate pushdown and column pruning tests

- Implement ClickHouseTableProvider for format-based API access
- Add DataSourceRegister service file for 'clickhouse' format
- Create comprehensive test suite with query log verification:
  * Predicate pushdown verification via system.query_log
  * Column pruning verification via system.query_log
  * Schema inference with auto-generated ORDER BY
  * Nullable key support with allow_nullable_key setting
  * Write modes (Append, ErrorIfExists)
  * Null value handling
- Refactor getTable method into clean helper methods
- Update ExprUtils and AstVisitor for expression handling

ShimonSte committed:
Fix Spark 3.3 xxHash64 partitioning for SharedMergeTree

For SharedMergeTree tables, filter out unsupported partition expressions
(like xxHash64 in Spark 3.3) while keeping supported ones. This maintains
Spark-side partitioning performance benefits while avoiding AnalysisException
errors.

- Keep xxHash64 commented out in Spark 3.3 ExprUtils (V2ExpressionUtils.toCatalyst limitation)
- Add SharedMergeTree detection and partition expression filtering
- Applied to all Spark versions (3.3, 3.4, 3.5, 4.0)
- Add Apache License header to DataSourceRegister service file

ShimonSte committed:
Implement SaveMode.Overwrite support for Spark 4.0

- Remove SupportsOverwrite from ClickHouseTable (only WriteBuilder should implement it)
- Implement SupportsOverwrite in ClickHouseWriteBuilder with overwrite mode tracking
- Add truncation logic in ClickHouseBatchWrite that executes TRUNCATE TABLE before writing
- Simplify distributed table truncation to use ON CLUSTER once instead of looping
- Handle AlwaysTrue filter to distinguish full table overwrite from partition filters
- Add overwrite test verifying old data is replaced with new data
- Add OVERWRITE_BY_FILTER capability to support overwrite operations

Note: OVERWRITE_DYNAMIC capability not included as partition-specific overwrite
is not yet implemented - currently only full table truncate is supported.

mzitnik commented on spark-4.0/clickhouse-spark-it/src/test/scala/org/apache/spark/sql/clickhouse/single/ClickHouseTableProviderSuite.scala:
```diff
@@ -0,0 +1,514 @@
+    }
+  }
+
+  test("schema inference works correctly") {
```
For schema inference, we can add more types

mzitnik reviewed (commented):
Are we also testing TableProvider in cluster mode (ClickHouse Cluster)?

ShimonSte committed:
Fix Jackson version conflict for Databricks compatibility

- Replace findAndRegisterModules() with explicit DefaultScalaModule registration
- Prevents conflict with Databricks' shaded Hudi Jackson modules (v2.15.x)
- Fixes JsonMappingException when using Spark 4.0 connector in Databricks
- Ensures compatibility with Spark 4.0's Jackson 2.18.2

ShimonSte committed:
Register JSR310 module for Java 8 date/time support

- Add JavaTimeModule registration for LocalDateTime serialization
- Fixes CI test failures: 'Java 8 date/time type LocalDateTime not supported'
- Maintains Databricks compatibility by avoiding findAndRegisterModules()
- Manually registers only required modules: DefaultScalaModule + JavaTimeModule

ShimonSte committed:
Fix Jackson module registration with graceful fallback

- Use try-catch to attempt findAndRegisterModules() first
- Falls back to manual registration (Scala + JSR310) on failure
- Handles Databricks shaded Hudi Jackson conflicts gracefully
- Scalable: auto-discovers new modules in normal environments
- Fixes CI test failures for LocalDateTime serialization

ShimonSte committed:
Add TableProvider implementation for Spark 3.3, 3.4, and 3.5

- Port ClickHouseTableProvider from Spark 4.0 to earlier versions
- Add DataSourceRegister service files for 'clickhouse' format
- Add comprehensive test suite with:
  * Format API read/write tests
  * SaveMode tests (Append, ErrorIfExists)
  * Predicate pushdown verification via query_log
  * Column pruning verification via query_log
  * Schema inference with auto-generated ORDER BY
  * Nullable key support with allow_nullable_key setting
  * Null value handling
- Use JavaConverters for Scala 2.12 compatibility
- Revert JsonProtocol to simple findAndRegisterModules()

This enables format-based access pattern for all Spark versions:
  spark.read.format("clickhouse")
    .option("host", "localhost")
    .option("database", "default")
    .option("table", "my_table")
    .load()

ShimonSte committed:
Add SaveMode.Overwrite support for Spark 3.3, 3.4, and 3.5

- Implement SupportsOverwrite in ClickHouseWriteBuilder with overwrite mode tracking
- Add truncation logic in ClickHouseBatchWrite that executes TRUNCATE TABLE before writing
- Simplify distributed table truncation to use ON CLUSTER once instead of looping
- Handle AlwaysTrue filter to distinguish full table overwrite from partition filters
- Add OVERWRITE_BY_FILTER capability to ClickHouseTable
- Add overwrite test verifying old data is replaced with new data
- Remove distributionStrictlyRequired (Spark 4.0 only API)

This enables SaveMode.Overwrite for all Spark versions:
  df.write.format("clickhouse")
    .mode(SaveMode.Overwrite)
    .option("host", "localhost")
    .option("table", "my_table")
    .save()

ShimonSte committed:
Add Apache License header to DataSourceRegister service files

- Add license header to Spark 3.3, 3.4, and 3.5 DataSourceRegister files
- Spark 4.0 already had the license header

ShimonSte committed:
Fix withSimpleTable function signature to pass parameters

The withSimpleTable function was not passing actualDb and tbl parameters
to the function block, causing tests to fail. This was a critical bug where:

- Function signature was: (f: => Unit) instead of (f: (String, String) => Unit)
- Function call was: f instead of f(actualDb, tbl)
- All test usages expect: { (actualDb: String, actualTbl: String) => ... }

This fix aligns Spark 3.3, 3.4, and 3.5 with the Spark 4.0 implementation
which already had the correct signature.

ShimonSte committed:
Fix predicate pushdown test to filter query log more specifically

The test was failing because the query log was capturing queries from
multiple filter operations within the 10-second window. The fix adds:

1. Additional LIKE '%name%' filter to narrow down to queries with name column
2. Support for both `name` = 'Bob' and `name`='Bob' formats (with/without spaces)

This ensures the test correctly identifies the specific query with the
name filter instead of getting confused by earlier queries with value filters.

ShimonSte committed:
Add Spark 4.0 fix and simplify predicate pushdown test

- Add LIKE '%name%' filter to Spark 4.0 test
- Remove redundant check for `name`='Bob' (without spaces) in all versions
- ClickHouse query log format is consistent, so single check is sufficient

mzitnik reviewed (changes requested):
General 
- Let's go over all logging levels and decide the correct level and if we need to log 
- the first comment on extending the coverage 
- Did we solve the conflict with Databricks on Spark 4.0

ShimonSte committed:
Add TableProvider support with Databricks compatibility

- Implement TableProvider API for all Spark versions (3.3, 3.4, 3.5, 4.0)
- Add shaded Jackson in Spark 4.0 runtime jar to avoid Databricks Hudi conflicts
- Add readTableWithBothAPIs helper to test both Catalog and TableProvider APIs
- Update logging levels and remove redundant log statements
- Remove Databricks-specific jar variant, use single jar with shaded Jackson
- Remove addDefaultOrderBy and require explicit order_by for table creation
- Add validation to ensure order_by columns exist in schema
- Fix import conflicts and rebase issues

pan3793 commented on deploy.gradle:
```diff
@@ -81,11 +81,22 @@ subprojects {
                         artifact javadocJar
+                        
+                        // For databricks variant, set the artifact ID to match the normal runtime (without -databricks)
+                        if (project.name.contains("-databricks")) {
```
Can you guys explain why we need to build a variant for Databricks Spark?

pan3793 replied:
OK, I guess Databricks Spark may use a different version of Jackson. In this case, I think we can always shade the Jackson in the runtime jar instead of produce a new variant.

ShimonSte replied:
We were planning to have two variants one shaded (for databricks) and one not (I am not strongly apposed to have only one shaded version, our rational was not to force users with a shaded jar when they don't need it).
We have managed to reach out to databricks yesterday, they will check if the shading issue could be fixed for us.
I'll changed the PR to draft for several days to give them a chance to fix it.

pan3793 replied:
Class shading and relocation are quite common practices in Java projects. I lean towards having a unified runtime jar that includes the relocated Jackson classes, if it's necessary.

pan3793 commented on spark-4.0/build.gradle:
```diff
@@ -44,16 +44,27 @@ project(":clickhouse-spark-runtime-${spark_binary_version}_$scala_binary_version
+        
+        // Exclude Jackson service files from external sources to prevent auto-discovery
+        // of incompatible modules (e.g., from Databricks Hudi)
+        exclude "META-INF/services/com.fasterxml.jackson.databind.Module"
```
you should relocate it instead of excluding it ...

now sure if the solution provided by https://github.com/GradleUp/shadow/issues/105 still works, as gradle is famous for its breaking changes.

pan3793 commented on spark-4.0/build.gradle:
```diff
@@ -44,16 +44,27 @@ project(":clickhouse-spark-runtime-${spark_binary_version}_$scala_binary_version
+        // of incompatible modules (e.g., from Databricks Hudi)
+        exclude "META-INF/services/com.fasterxml.jackson.databind.Module"
+        
         mergeServiceFiles()
```
oh wait, it is already here, then the runtime jar should include a relocated service manifest, why is it still in conflict with hudi?

pan3793 commented on deploy.gradle:
```diff
@@ -84,8 +84,8 @@ subprojects {
                             name = "Spark ClickHouse Connector"
-                            url = "https://github.com/clickhouse/spark-clickhouse-connector"
                             description = "Spark ClickHouse Connector build on Apache Spark DataSourceV2 API."
+                            url = "https://github.com/clickhouse/spark-clickhouse-connector"
```
please revert unnecessary changes.

pan3793 commented on spark-3.3/clickhouse-spark/src/main/scala/com/clickhouse/spark/ClickHouseTableProvider.scala:
```diff
@@ -0,0 +1,215 @@
+        )
+      }
+
+    log.info(s"Auto-generated ORDER BY clause: $orderByColumn")
```
I'm not sure if such an automatic behavior is a good idea, could you please elaborate more about your consideration? and this at least requires a config to allow the user to disable it.

ShimonSte committed:
Add logging for batchSize and format in read/write operations

- Add logging when creating ClickHouse writers with format and batchSize
- Add logging when creating ClickHouse readers with format
- Add logging when flushing batches due to batchSize being reached
- Update log4j2.xml to output connector logs to console
- Update test to use batchSize=3122 for clearer verification
- Fix SSL and timezone option handling in ClickHouseTableProvider

ShimonSte committed:
Add timezone and ssl_mode support to createCatalog in Spark 3.3, 3.4, and 3.5

ShimonSte committed:
Fix Scala 2.12 compatibility: replace view.filterKeys with filter

ShimonSte committed:
Add missing distributionStrictlyRequired to Spark 3.4 and 3.5

mzitnik commented:
@ShimonSte, can you confirm that all our versions (Spark 3.3/3.4/3.5/4.0) are aligned with the same code
